File: backend/app/models/plan.py
from typing import Any, Literal, Optional

from pydantic import BaseModel


class PlanPassport(BaseModel):
    selected_variant: str
    planning_horizon_months: int
    checkpoint_count: int
    risk_count: int
    constraints_in_attention_count: int
    execution_progress: int


class PlanRisk(BaseModel):
    risk: str
    impact: Literal["Низкий", "Средний", "Высокий"]


class ConstraintInAttention(BaseModel):
    constraint: str
    actual_value: str
    impact: str


class PlanResponse(BaseModel):
    status: Literal["PENDING", "IN_PROGRESS", "COMPLETED", "FAILED"]
    selected_scenario_type: Optional[Literal["BALANCED", "CONSERVATIVE", "RISKY"]] = None
    plan_passport: Optional[PlanPassport] = None
    stages: list[Any] = []
    resource_loading_by_stages: list[Any] = []
    plan_risks: list[PlanRisk] = []
    constraints_in_attention: list[ConstraintInAttention] = []

File: frontend/src/app/api/plan.ts
import { api } from './client';

export type PlanStatus = 'NOT_STARTED' | 'IN_PROGRESS' | 'COMPLETED' | 'FAILED';

export type PlanCompositionItem = {
  project_name: string;
  share: number;
  roles: string;
};

export type PlanResourceItem = {
  role: string;
  required_hours: number;
  available_hours: number;
  loading_percent: number;
  loading_degree: string;
};

export type PlanCheckpoint = {
  name: string;
  date: string;
  status: string;
};

export type PlanStage = {
  stage_number: number;
  name: string;
  period: string;
  start_date: string;
  end_date: string;
  stage_status: string;
  checkpoint: PlanCheckpoint;
  composition: PlanCompositionItem[];
  resources: PlanResourceItem[];
};

export type PlanPassport = {
  selected_variant: string;
  planning_horizon_months: number;
  checkpoint_count: number;
  risk_count: number;
  constraints_in_attention_count: number;
  execution_progress: number;
};

export type PlanRisk = {
  risk: string;
  impact: 'Высокий' | 'Средний' | 'Низкий';
};

export type ConstraintInAttention = {
  constraint: string;
  actual_value: string;
  impact: string;
};

export type ResourceLoadingByStage = {
  stage_number: number;
  analysts_percent: number;
  developers_percent: number;
  testers_percent: number;
};

export type PlanData = {
  plan_passport: PlanPassport;
  stages: PlanStage[];
  resource_loading_by_stages: ResourceLoadingByStage[];
  plan_risks: PlanRisk[];
  constraints_in_attention: ConstraintInAttention[];
};

export type PlanResponse = {
  status: PlanStatus;
  plan: PlanData | null;
  error?: string | null;
};

export const planApi = {
  get: (contextId: string) => api.get<PlanResponse>(`/contexts/${contextId}/plan`),
};
